### PR TITLE
o/snapstate: autorefresh phase1 for refresh-control

### PR DESCRIPTION
**Pull request by stolowski**
Implement autorefresh method that initiates auto-refresh: updates refresh-candidates, creates hooks and a new gating task ("conditional-auto-refresh"). Phase 2 will be handled by doAutoRefreshGate.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1818,7 +1818,90 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state.TaskSet, error) {
+	user, err := userFromUserID(st, 0)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	refreshOpts := &store.RefreshOptions{IsAutoRefresh: true}
+	updates, snapstateByInstance, ignoreValidationByInstanceName, err := refreshCandidates(ctx, st, nil, user, refreshOpts)
+	if err != nil {
+		// XXX: should we reset "refresh-candidates" to nil in state for some types
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1846]: I principle not, but it depends how exactly we will implement invalidation/resetting of them

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1818,7 +1818,90 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+	hints, err := refreshHintsFromCandidates(st, updates, ignoreValidationByInstanceName, deviceCtx)
+	if err != nil {
+		return nil, nil, err
+	}
+	st.Set("refresh-candidates", hints)
+
+	if len(updates) == 0 {
+		return nil, nil, nil
+	}
+
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1863]: we need to do filtering based on conflicts at least as mushc as UpdateMany would have done here

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1818,7 +1818,90 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+	var hooks *state.TaskSet
+	if len(affectedSnaps) > 0 {
+		hooks = createGateAutoRefreshHooks(st, affectedSnaps)
+	}
+
+	// gate-auto-refresh hooks, followed by conditional-auto-refresh task waiting
+	// for all hooks.
+	ar := st.NewTask("conditional-auto-refresh", "Run select refreshes")
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1889]: Run auto-refresh for ready snaps ? or something like that ? not sure myself about ready

**Review by pedronis — COMMENTED** [review]
some comments

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go:1889]: Done (yeah, we may want to review this name anyway).

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go:1863]: Added.

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -1818,7 +1818,106 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+	if err != nil {
+		// XXX: should we reset "refresh-candidates" to nil in state for some types
+		// of errors?
+		return nil, nil, err
+	}
+	deviceCtx, err := DeviceCtxFromState(st, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	hints, err := refreshHintsFromCandidates(st, candidates, ignoreValidationByInstanceName, deviceCtx)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/snapstate/snapstate.go:1854]: There was no particular reason for the list, in hindsight I agree map makes more sense, will change it, thanks.

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -1818,7 +1818,106 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+	// check conflicts
+	fromChange := ""
+	for _, up := range candidates {
+		snapst := snapstateByInstance[up.InstanceName()]
+		if err := checkChangeConflictIgnoringOneChange(st, up.InstanceName(), snapst, fromChange); err != nil {
+			logger.Noticef("cannot refresh snap %q: %v", up.InstanceName(), err)
+		} else {
+			updates = append(updates, up)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/snapstate/snapstate.go:1876]: Added

File: overlord/snapstate/refreshhints.go [hunk 6]
```diff
@@ -149,7 +149,7 @@ func refreshHintsFromCandidates(st *state.State, updates []*snap.Info, ignoreVal
 		}
 	}
 
-	hints := []*refreshCandidate{}
+	hints := make(map[string]*refreshCandidate)
```
Comment (pedronis) [inline overlord/snapstate/refreshhints.go:152]: should we use len(updates) here, it seems a good guess even if there's a continue below?

**Review by pedronis — APPROVED** [review]
thanks, one small comment

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -1818,7 +1818,111 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
+	var updates []*snap.Info
+
+	// check conflicts
+	fromChange := ""
+	for _, up := range candidates {
+		snapst := snapstateByInstance[up.InstanceName()]
+		if err := checkChangeConflictIgnoringOneChange(st, up.InstanceName(), snapst, fromChange); err != nil {
+			logger.Noticef("cannot refresh snap %q: %v", up.InstanceName(), err)
+		} else {
+			// all snaps in updates are now considered to be operated on and should
```
Comment (bboozzoo) [inline overlord/snapstate/snapstate.go:1870]: Shouldn't this comment be right after the `for{}` loop?

File: overlord/snapstate/autorefresh_gating_test.go [hunk 8]
```diff
@@ -524,3 +561,223 @@ func (s *autorefreshGatingSuite) TestCreateAutoRefreshGateHooks(c *C) {
...
+
+	// enable gate-auto-refresh-hook feature
+	tr := config.NewTransaction(s.state)
+	tr.Set("core", "experimental.gate-auto-refresh-hook", true)
+	tr.Commit()
+
+	_, tss, err = snapstate.AutoRefresh(context.TODO(), st)
+	c.Check(err, IsNil)
+	c.Assert(tss, HasLen, 2)
+	c.Check(tss[0].Tasks()[0].Kind(), Equals, "conditional-auto-refresh")
```
Comment (bboozzoo) [inline overlord/snapstate/autorefresh_gating_test.go:606]:
```
// TODO: verify conditional-auto-refresh task data
```

**Review by bboozzoo — APPROVED** [review]
LGTM, smallish tweaks

**Pull request merged**